make reflector config public

azdle opened the pull request:
There's a few things in this that I'm not sure are the right way to do things. Inline comments coming up in a sec.

This is a draft both because I want to get some specific feedback before I feel good about merging and I'd like either Jon or I to actually integrate it into something before I merge.

azdle commented on client-libraries/rust/modality-ingest-client/src/types.rs:
```diff
@@ -265,6 +312,32 @@ impl Encode for AttrVal {
+        } else if let Ok(v) = s.parse::<Uuid>() {
+            v.into()
+        } else {
+            AttrVal::String(s.trim_matches('"').into())
```
Right now, I'm accepting both `foo` and `"foo"` as an `AttrVal::String`. My original assumption was that anything that didn't parse as something else should parse as a string, but I'm not sure if that's really what we want to go with. I'm absolutely of two minds about it, so I'd appreciate a tie-break opinion.

1. Only accept strings that are wrapped in `"`.
2. Accept strings that are wrapped in `"` (good for ensuring things aren't parsed as something else), but also interpret anything else not otherwise parsed as a string.

ZackPierce replied:
This is a toughy. Thinking...

azdle replied:
To add a bit more context. This is only for the `FromStr` impl, which is essentially only used for `my_str.parse()`. We also have `From<&str>` & `From<String>` which are independent. So this is only deciding for the case where we're trying to "Parse this unknown `AttrVal` from a string" and not "I have something that is explicitly a string that should `.into()` an `AttrVal::String`.

I think that makes the "require quotes" option viable. I'm just not sure if we like the idea of errororing out while parsing the config file or crossing fingers in ambiguous cases more.

ZackPierce replied:
I think we want to do option 2 here. Probably ought to trim both `"` and `'`, in truth.

azdle commented on client-libraries/rust/modality-ingest-client/src/types.rs:
```diff
@@ -300,3 +373,102 @@ impl From<bool> for AttrVal {
+    }
+}
+
+// TODO: Should I actually assume that any UUID is a timeline id?
+impl From<Uuid> for AttrVal {
+    fn from(u: Uuid) -> AttrVal {
+        AttrVal::TimelineId(Box::new(u.into()))
+    }
+}
```
Should I automatically assume this? This probably depends on the answer to how we parse strings.

ZackPierce replied:
Seems fine for now.

azdle commented on client-libraries/rust/modality-reflector-config/src/lib.rs:
```diff
@@ -0,0 +1,990 @@
+    use std::time::Duration;
+    use url::Url;
+
+    /// A local string-ish AttrKey value.
+    #[derive(Clone, Debug, PartialEq, Eq, PartialOrd)]
+    pub struct AttrKey(String);
+
+    impl AttrKey {
+        fn new(k: String) -> Self {
+            Self(k)
+        }
+    }
+
+    impl std::fmt::Display for AttrKey {
+        fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> Result<(), std::fmt::Error> {
+            write!(f, "{}", self.0)
+        }
+    }
```
I added a different `AttrKey` that is string-ish rather than int-ish like it is in the SDK. Is it confusing that they're called the same thing even though they're in different places. Do we already have names to distinguish between string-ish AttrKey and int-ish AttrKey anywhere else?

ZackPierce replied:
The int-ish `AttrKey` is an implementation detail - a caching technique specific to the `CBOR` ingest protocol.  Re: name differences, crate-and-mod-path differences is distinguishing enough for me, but your mileage may vary. 

If we need a type alias for the CBOR one, then it's probably fine to invent one.

azdle replied:
Okay. In that case I'm not going to try to string-ish-ify the ingest client `AttrKey` right not since that'd be either a breaking change or take some tricksy doing to not be breaking, I'll just leave this with the two different module paths.

ZackPierce replied:
Now that I notice you're pulling in the whole ingest crate to avoid duplicating the attribute-value-related types, I can see why this would be a point of confusion.

ZackPierce replied:
Perhaps a simple and moderately clarifying thing to do would be to move the protocol-specific `modality_ingest_client::types::AttrKey` to be at `modality_ingest_client::protocol::AttrKey` instead.

ZackPierce replied:
You're also making the right choice of not string-ifying the ingest client's types further at this time, I think.

azdle commented on client-libraries/rust/modality-ingest-client/src/types.rs:
```diff
@@ -300,3 +373,102 @@ impl From<bool> for AttrVal {
+            ))),
+            "bec14bc0-1dea-4b68-b138-62f7b6827e35".parse()
+        );
+
+        // Timestamp
+        // N.B. This is impossible to parse as an `AttrVal` since it's just a number. Could try
+        // parsing more complex date strings?
+
```
I'm assuming Timestamp's inner Nanoseconds is the format we would use in a config file. Is that a bad assumption? Should I try parsing some unambiguous datetime formats to get this? Is there some other format we want to use that wouldn't first be parsed as just an int?

ZackPierce replied:
Re: Nanoseconds, I think that's probably a good enough assumption.
I don't think we need to add date string parsing now.

Honestly, I think that if we wanted to make more use of the Timestamp variant now, the only approach we have available is to peek over at the relevant `AttrKey` and see if it's one of the designated attributes semantically associated with timey-values. (e.g. `event.timestamp` and `event.interaction.remote_timestamp`).

azdle replied:
:+1: Okay, I won't change anything here.

> Honestly, I think that if we wanted to make more use of the Timestamp variant now, the only approach we have available is to peek over at the relevant AttrKey and see if it's one of the designated attributes semantically associated with timey-values. (e.g. event.timestamp and event.interaction.remote_timestamp).

Totally, getting a "well known attr key" -> AttrVal type mapping is something that's been on my list to get into the SDK since before it was published.

ZackPierce commented on client-libraries/rust/modality-ingest-client/src/types.rs:
```diff
@@ -265,6 +312,32 @@ impl Encode for AttrVal {
+        // as our AttrVals support more interesting variants.
+        Ok(if let Ok(v) = s.to_lowercase().parse::<bool>() {
+            v.into()
+        } else if let Ok(v) = s.parse::<i64>() {
```
There's no need for an `i64` parsing branch, the i128 parsing and AttrVal converting branch will shrink to fit i64 internally if possible.

ZackPierce commented on client-libraries/rust/modality-ingest-client/src/types.rs:
```diff
@@ -265,6 +312,32 @@ impl Encode for AttrVal {
+            v.into()
+        } else if let Ok(v) = s.parse::<f64>() {
+            v.into()
+        } else if let Ok(v) = s.parse::<Nanoseconds>() {
```
Currently this isn't going to match anything not already caught by `i128`. Can take out to avoid wasted cycles.

ZackPierce commented on client-libraries/rust/modality-reflector-config/Cargo.toml:
```diff
@@ -0,0 +1,14 @@
+edition = "2021"
+
+[dependencies]
+modality-ingest-client = { version = "0.1", path = "../modality-ingest-client" }
```
It's probably a bit much to pull in the entire ingest crate just for `AttrVal` representation and parsing, especially since it has relatively heavy dependencies on `tokio` and the TLS stuff.  A quick fix would be to make the modality-ingest-client crate have those dependencies be optional, tied to a "client" feature and conditionally compile the `client` module based on that feature. Said feature could be part of the default feature-set, but `modality-reflector-config` could opt out of the default features.

azdle replied:
I wondered about that.

I justified doing this to myself by assuming that anything is going to be run by the reflector is also going to talk to the reflector and would essentially need this library anyway.

Is that actually a reasonable assumption?

ZackPierce replied:
Config inspection or editing tooling might want to read config without running an ingest client.

Separately, the mutation protocol configuration is also managed in the reflector config.  When the mutation story advances further, folks will probably want to be able to read the config for that side of the world without pulling in the ingest client guts also.

azdle commented:
@ZackPierce I think I've addressed all your comments so far. And since Jon has integrated this into his thing I'm taking this out of draft.

I ended up just doing the splitting of the `types` (and also the `protocol`) module out into a new `modality-api` crate now instead of feature flagging the client for now. It was ~2% more work and felt way less nonsensical than feature flagging the client out of the client.

ZackPierce commented:
Looking better.

I think the `modality-api` crate should mostly be AttrKey, AttrVal and its variants. The rest of the ingest-protocol specific types (e.g. `IngestResponse`) should be in the ingest client crate.

azdle commented:
@ZackPierce I've moved out as much of the protocol stuff from the `api` crate into a new `modality-ingest-protocol` crate. What remains in the `api` crate can't be moved without violating the orphan rule.

ZackPierce reviewed (approved):
:+1:

azdle commented:
I want to get a PR up for handling the client's breaking changes in `tracing-modality` before we cut a release for this. Will put up a release PR once I have that ready.
